modules/openapi-generator/src/main/resources/swift5/libraries/urlsession/URLSessionImplementations.mustache

```diff
@@ -52,7 +52,22 @@ private var urlSessionStore = SynchronizedDictionary<String, URLSession>()
         let sessionDelegate = SessionDelegate()
         sessionDelegate.credential = credential
         sessionDelegate.taskDidReceiveChallenge = taskDidReceiveChallenge
-        return URLSession(configuration: configuration, delegate: sessionDelegate, delegateQueue: nil)
+
+        var hasher = Hasher()
+        hasher.combine(configuration)
+        hasher.combine(credential)
+        hasher.combine(sessionDelegate)
+
+        let hash = hasher.finalize()
+
+        if let session = urlSessionStore[hash] {
+            return session
```

Comment:
@DavidHorvath I was testing this locally, but it's not reusing sessions.
I set a breakpoint here, but it never hit the breakpoint.
The problem seems to be related to `hasher.combine(sessionDelegate)`

Reply:
@4brunu Hi, yeah you are right. I've tried it before adding the sessionDelegate to the Hasher. 
Sorry about that. I will update my PR

Reply:
@4brunu Updated it

Reply:
@4brunu But I need to update the sample codes as well so just another minute

Reply:
No problem 🙂 I'm just wondering if this is safe or not without the SessionDelegate.
Imagine the following scenario.
The user makes a request to the endpoint A.
Then he repetes the same request.
The second request will invoque the first delegate, not the second one.
Or am I thinking wrong?

Reply:
I think we won't have any issue with this, because we only have one `requestBuilderFactory` and in that one `sessionDelegate` which is implemented privately in the `URLSessionImplementations` as well. 
So it won't change

Reply:
Meanwhile I've updated my PR with the sample codes as well

Reply:
We only have one `requestBuilderFactory`, but the `requestBuilderFactory` generates a new instance of `URLSessionRequestBuilder` and `URLSessionDecodableRequestBuilder` for each request, which means that each request we have a different `sessionDelegate`.

Reply:
I see. Let me think about a solution

Reply:
@4brunu 

So right now
- `RequestBuilderFactory` creates a new instance of `URLSessionRequestBuilder` and `URLSessionDecodableRequestBuilder` for every request
- Every time a new `URLSessionRequestBuilder` is created for a request, it creates a new `URLSession` and a new `SessionDelegate)` which handles the request challenges

But if we don’t want to create a new URLSession for each request
- We can change the SessionDelegate(URLSessionDelegate) to a TaskDelegate(URLSessionTaskDelegate) and just use that for every request

**If we want to handle credential requests in a session-level we would need to implement (URLSessionDelegate)**
urlSession(_:didReceive:completionHandler:)
https://developer.apple.com/documentation/foundation/urlsessiondelegate/1409308-urlsession

**The implementation we have currently is for requests (URLSessionTaskDelegate)**
urlSession(_:task:didReceive:completionHandler:)
https://developer.apple.com/documentation/foundation/urlsessiontaskdelegate/1411595-urlsession

https://developer.apple.com/documentation/foundation/url_loading_system/handling_an_authentication_challenge

What do you think? If you want I can add a new commit for you to check what I'm thinking about

One thing is not clear for me, what would be the preferred way to assign `taskDidReceiveChallenge` & `taskCompletionShouldRetry`? 
Can you please clarify that for me?
Thanks

Reply:
Hi @DavidHorvath,

> One thing is not clear for me, what would be the preferred way to assign taskDidReceiveChallenge & taskCompletionShouldRetry?


I would say maybe in the init method of the subclass of  `RequestBuilder`.

For example here, if there was an init method.


https://github.com/OpenAPITools/openapi-generator/blob/0bb4f186ec7e776efbe370be89798ed2f659c9ff/samples/client/petstore/swift5/urlsessionLibrary/SwaggerClientTests/SwaggerClient/BearerDecodableRequestBuilder.swift#L22


I'm not to worried about `taskCompletionShouldRetry`.

My main concern is with `taskDidReceiveChallenge` and the credential handle, that could be address with your solution of replacing the SessionDelegate.

If you could add a commit to show your idea of replacing the `URLSessionDelegate` with the `URLSessionTaskDelegate` that would be nice.
I think that maybe the could be something on the right track, but looking at the concrete implementation of it would make things easier.

Thanks for this PR, this is a nice improvement 👍

Reply:
@4brunu I've added 2 new commits:

- Add separate TaskDelegate as data task's delegate
- Store URLSession's delegate to keep reference

**Handle Task/Request Challenge**

1. Renamed `SessionDelegate` to `TaskDelegate`. The current delegate implementation `urlSession(_:task:didReceive:completionHandler: )` is in the [URLSessionTaskDelegate](https://developer.apple.com/documentation/foundation/urlsessiontaskdelegate/1411595-urlsession)
so I changed the delegates to `URLSessionTaskDelegate` as well. (This is inherited from `URLSessionDelegate` so we can use the same TaskDelegate for the URLSession later as well)
2. Instantiate a new private TaskDelegate for the URLSessionRequestBuilder, set the credentials and taskDidReceiveChallenge and set it as the `dataTask.delegate` so we can handle the challenge for each request.

**Handle Session Challenge**

- Later if we want/need, we can have a separate SessionDelegate class which implements URLSessionDelegate and [ urlSession(_:didReceive:completionHandler:)](https://developer.apple.com/documentation/foundation/urlsessiondelegate/1409308-urlsession) to handle session level challenges
- **The documentation says:** `If you do not implement this method, the session calls its delegate’s urlSession(_:task:didReceive:completionHandler:) method instead.` 
- So to handle session level challenges (until we implement a separate SessionDelegate) we can use the TaskDelegate for the URLSession as well

So what I did:
1. I've created a new `sessionDelegateStore` to store the delegates for the URLSessions
2. If we don't have a stored URLSession, I create a new TaskDelegate instance for the new URLSession, set the credentials and taskDidReceiveChallenge (as before) and store it in the sessionDelegateStore

Questions/Notes:
- Is it okay that I renamed `SessionDelegate` for `TaskDelegate` even if we use it for the URLSession as well? Basically it is a task delegate but as the documentation says if we don't have the session delegate implementation it will use the task delegate to handle the challenges
- Currently both the URLSession's challenge and the DataTask's challenge will call the same `taskDidReceiveChallenge`. If we want to have a separate `SessionDelegate` we will need another new closure like `sessionDidReceiveChallenge`

What do you think about it? Please let me know if I missed something

Reply:
Hey,

Overall this looks good to me, setting the delegate on the `URLSessionTask` is a great idea.
But I found one big problem, the `delegate` of `URLSessionTask` is only available on iOS 15 an above.

Looking at `URLSessionTask` gave me another idea.
The `URLSessionTask` has a property `taskIdentifier`.
What if we have only one `URLSession` and only one `URLSessionDelegate` reusable for all the requests.
Inside the `URLSessionDelegate` instead of having only one `taskDidReceiveChallenge`, we would have a dictionary of them, and the `URLSessionTask.taskIdentifier` would be the key for it.
Then, when the method `func urlSession(_ session: URLSession, task: URLSessionTask, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)` get's called, we would get the callback using the `task.taskIdentifier`.

What do you think of this?
Or do you have a better idea to workaround the iOS 15 limitation?

Reply:
😕 Yeah, you are right. I missed that this is from iOS15 only...

I think your solution can work. I will try it out

Reply:
@4brunu 

I just added a new commit. Please check its okay.
If you think this is the right way I will squash my commits for the changes and generate new sample codes

Reply:
Also, what do you think about adding a typealias for the closure type to be more readable?
`public typealias {{projectName}}APIChallengeHandler = ((URLSession, URLSessionTask, URLAuthenticationChallenge) -> (URLSession.AuthChallengeDisposition, URLCredential?))`